urldae1c57e59_mustafaakin_izuna_mapeditor_tgz-pJ8-in_mustafaak_izuna_actions_ActionTestJ8/src/in/mustafaak/izuna/waves/set/E.java

```diff
@@ -0,0 +1,126 @@
+package in.mustafaak.izuna.waves.set;
+
+import in.mustafaak.izuna.meta.WaveInfo;
+import in.mustafaak.izuna.actions.*;
+import in.mustafaak.izuna.waves.CircularInvade;
+import in.mustafaak.izuna.waves.Invade;
+import in.mustafaak.izuna.waves.RowColEnter;
+
+public class E {
+	static Action l = new Loop();
+
+	public static WaveInfo w1() {
+		WaveInfo w = new WaveInfo();
+		Action s = new Start(-200, 100);
+		Action m = new Move(300, 300, 2000);
+		Action z = new ZigZag(200, 500, true, 1000);
+
+		for (int i = 0; i < 5; i++) {
+			w.putEnemy("a4", s, l, new Delay(i * 500 + 1), m, z, z, z);
+		}
+		return w;
+	}
+
+	public static WaveInfo w2() {
+		WaveInfo w = new WaveInfo();
+		Action s1 = new Start(-200, 100);
+		Action s2 = new Start(-200, 250);
+		Action m = new Move(300, 300, 2000);
+		Action z = new ZigZag(200, 500, true, 1000);
+
+		for (int i = 0; i < 2; i++) {
+			w.putEnemy("a4", s1, l, new Delay(i * 1000 + 1), m, z, z, z);
+			w.putEnemy("a4", s2, l, new Delay(i * 1000 + 1), m, z, z, z);
+		}
+
+		return w;
+	}
+
+	public static WaveInfo w3() {
+		WaveInfo w = new WaveInfo();
+		Action s1 = new Start(-200, 100);
+		Action m = new Move(300, 0, 1000);
+		Action r = new Rectangle(400, 200, true, 1000);
+
+		for (int i = 0; i < 6; i++) {
+			w.putEnemy(i % 2 == 0 ? "a3" : "a4", s1, new Delay(i * 500), m, l, r);
+		}
+
+		return w;
+	}
+
+	public static WaveInfo w4() {
+		WaveInfo w1 = new CircularInvade(3, -250, -200, 400, 400, 100, "a3", 5000);
+		WaveInfo w2 = new CircularInvade(3, 900, -200, -400, 400, 100, "a4", 5000);
+
+		return WaveInfo.combine(w1, w2);
+	}
+
+	public static WaveInfo w5() {
+		WaveInfo w1 = new CircularInvade(3, -250, -200, 400, 500, 70, "a3", 5000);
+		WaveInfo w2 = new CircularInvade(3, 900, -200, -400, 500, 70, "a3", 5000);
+
+		WaveInfo w3 = new RowColEnter(RowColEnter.FROM_UP, 150, 50, 1, 3, 150, 0, "a4", 2500);
+		Action.wiggle(w3, 0, 50, 2000);
+		return WaveInfo.combine(w1, w2, w3);
+	}
+
+	public static WaveInfo w6() {
+		WaveInfo w1 = new RowColEnter(RowColEnter.FROM_UP, 25, 50, 1, 5, 150, 0, "a1", 2500);
+		WaveInfo w2 = new RowColEnter(RowColEnter.FROM_UP, 100, 250, 1, 4, 150, 0, "a2", 2500);
+		WaveInfo w3 = new RowColEnter(RowColEnter.FROM_UP, 150, 450, 1, 3, 150, 0, "a3", 2500);
+		Action.wiggle(w1, 0, -50, 2000);
+		Action.wiggle(w2, 0, 50, 2000);
+		Action.wiggle(w3, 0, 100, 3000);
+
+		return WaveInfo.combine(w1, w2, w3);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 69:76, 80:87.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.